src/env.cpp

```diff
@@ -198,18 +198,54 @@ wcstring_list_t null_environment_t::get_names(int flags) const {
     return {};
 }
 
-/// Set up the USER variable.
-static void setup_user(bool force) {
+/// Set up the USER and HOME variable.
+static void setup_user() {
     auto &vars = env_stack_t::globals();
-    if (force || vars.get(L"USER").missing_or_empty()) {
-        struct passwd userinfo;
-        struct passwd *result;
-        char buf[8192];
-        int retval = getpwuid_r(getuid(), &userinfo, buf, sizeof(buf), &result);
+    auto uid = getuid();
+    auto user_var = vars.get(L"USER");
+    struct passwd userinfo;
+    struct passwd *result;
+    char buf[8192];
+
+    // If we have a $USER, we try to get the passwd entry for the name.
+    // If that has the same UID that we use, we assume the data is correct.
+    if (!user_var.missing_or_empty()) {
+        std::string unam_narrow = wcs2string(user_var->as_string());
+        int retval = getpwnam_r(unam_narrow.c_str(), &userinfo, buf, sizeof(buf), &result);
         if (!retval && result) {
-            const wcstring uname = str2wcstring(userinfo.pw_name);
-            vars.set_one(L"USER", ENV_GLOBAL | ENV_EXPORT, uname);
+            if (result->pw_uid == uid) {
+                // The uid matches but we still might need to set $HOME.
+                if (vars.get(L"HOME").missing_or_empty()) {
+                    if (userinfo.pw_dir) {
+                        const wcstring dir = str2wcstring(userinfo.pw_dir);
+                        vars.set_one(L"HOME", ENV_GLOBAL | ENV_EXPORT, dir);
+                    } else {
+                        vars.set_empty(L"HOME", ENV_GLOBAL | ENV_EXPORT);
+                    }
+                }
+                return;
+            }
+        }
+    }
+
+    // Either we didn't have a $USER or it had a different uid.
+    // We need to get the data *again* via the uid.
+    // TODO: Do we need euid() here instead?
+    int retval = getpwuid_r(uid, &userinfo, buf, sizeof(buf), &result);
+    if (!retval && result) {
+        const wcstring uname = str2wcstring(userinfo.pw_name);
+        vars.set_one(L"USER", ENV_GLOBAL | ENV_EXPORT, uname);
+        if (userinfo.pw_dir) {
+            const wcstring dir = str2wcstring(userinfo.pw_dir);
+            vars.set_one(L"HOME", ENV_GLOBAL | ENV_EXPORT, dir);
```

Comment:
Here I think the previous behaviour of only setting HOME when it is empty should be preserved. This is also not required as `su` (and others like `sudo`, `doas` etc.) always update the HOME even when `--login` is not passed.